thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/events/HolidayEventsLoader.java

```diff
@@ -0,0 +1,337 @@
+    } catch (Exception e) {
+      LOG.error("Can't create http transport with google api.", e);
+    }
+  }
+
+  private static final DAORegistry DAO_REGISTRY = DAORegistry.getInstance();
+
+  /**
+   * Start.
+   */
+  public void start() {
+    scheduledExecutorService.scheduleAtFixedRate(this, 0, runFrequency.getSize(), runFrequency.getUnit());
+  }
+
+  /**
+   * Shutdown.
+   */
+  public void shutdown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  /**
+   * Fetch holidays and save to ThirdEye database.
+   */
+  public void run() {
+    EventManager eventDAO = DAO_REGISTRY.getEventDAO();
+
+    long start = System.currentTimeMillis();
+    long end = start + 2592000000L; // to get holidays within a month
+
+    List<Event> newHolidays = getAllHolidays(start, end);
+
+    // A map from new holiday to a set of country codes that has the holiday
+    Map<HolidayEvent, Set<String>> newHolidayEventToCountryCodes = new HashMap<>();
+
+    // Group by holidays and put the country codes in the corresponding set
+    for (Event holiday : newHolidays) {
+      HolidayEvent holidayEvent =
+          new HolidayEvent(holiday.getSummary(), EventType.HOLIDAY.toString(), holiday.getStart().getDate().getValue(),
+              holiday.getEnd().getDate().getValue());
+
+      if (!newHolidayEventToCountryCodes.containsKey(holidayEvent)) {
+        newHolidayEventToCountryCodes.put(holidayEvent, new HashSet<String>());
+      }
+      newHolidayEventToCountryCodes.get(holidayEvent).add(getCountryCode(holiday));
+    }
+
+    // Get the existing holidays in the time range
+    List<EventDTO> existingEvents = eventDAO.findEventsBetweenTimeRange(EventType.HOLIDAY.toString(), start, end);
+    Map<HolidayEvent, EventDTO> existingHolidayEventToEventDTO = new HashMap<>();
+    Map<HolidayEvent, Set<String>> existingHolidayEventToCountryCodes = new HashMap<>();
+
+    // Map existing event DTO to the tuple of HolidayEvent and country code set
+    for (EventDTO existingEventDTO : existingEvents) {
+      HolidayEvent holidayEvent = new HolidayEvent(existingEventDTO);
+      if (!existingHolidayEventToCountryCodes.containsKey(holidayEvent)) {
+        existingHolidayEventToCountryCodes.put(holidayEvent, new HashSet<String>());
+      }
+      if (existingEventDTO.getTargetDimensionMap() != null
+          && existingEventDTO.getTargetDimensionMap().get("countryCode") != null) {
+        existingHolidayEventToCountryCodes.get(holidayEvent)
+            .addAll(existingEventDTO.getTargetDimensionMap().get("countryCode"));
+      }
+      existingHolidayEventToEventDTO.put(holidayEvent, existingEventDTO);
+    }
+
+    for (Map.Entry<HolidayEvent, Set<String>> entry : newHolidayEventToCountryCodes.entrySet()) {
+      HolidayEvent newHolidayEvent = entry.getKey();
+      Set<String> newCountryCodes = entry.getValue();
+
+      if (existingHolidayEventToCountryCodes.containsKey(newHolidayEvent)) {
```

Comment:
as before, when the holiday changes dates, it is not picked up here. Ultimately, there are three cases: (1) new event appears, (2) existing event changes, and (3) event disappears. The code handles (1) and country additions for (2). I would modify this section to simply override any pre-existing events (and their country codes) with the same name for (2). For 3 you may need to introduce some kind of time-based cutoff and then simply delete events that aren't returned from the API.

Probably makes sense to wrap this logic into a testable unit, too.

Reply:
Thanks! I will add this logic and tests.

Reply:
Can we just remove all existing events within the specific time range and write the new ones?

Reply:
yes, that could work. this assumes that there's strict time barrier for moving around events, i.e. events from the future cannot move to the past and vice versa.

there may be an issue with RCA however - saved investigations keep track of event by id, and these events could be in the future (just a few days). If we delete events and the re-create them, they get a new id.

If we put the time barrier for delete/re-create in the future, e.g. by a week, this should work. It may, however, have some non-obvious side-effects.